src/main/java/sparksoniq/spark/iterator/flowr/GroupByClauseSparkIterator.java

```diff
@@ -209,17 +219,194 @@ private void linearizeTuples(List<FlworTuple> keyTuplePairs) {
         _rdd = this._child.getRDD(context);
         //map to pairs - ArrayItem [sort keys] , tuples
         JavaPairRDD<FlworKey, FlworTuple> keyTuplePair = this._rdd
-                .mapToPair(new GroupByToPairMapClosure(_variables));
+                .mapToPair(new GroupByToPairMapClosure(_expressions));
         //group by key
         JavaPairRDD<FlworKey, Iterable<FlworTuple>> groupedPair =
                 keyTuplePair.groupByKey();
         //linearize iterable tuples into arrays
-        this._rdd = groupedPair.map(new GroupByLinearizeTupleClosure(_variables));
+        this._rdd = groupedPair.map(new GroupByLinearizeTupleClosure(_expressions));
         return _rdd;
     }
 
     @Override
     public Dataset<Row> getDataFrame(DynamicContext context) {
-        return null;
+        if (this._child == null) {
+            throw new SparksoniqRuntimeException("Invalid groupby clause.");
+        }
+        Dataset<Row> df = _child.getDataFrame(context);
+        StructType originalInputSchema = df.schema();
+        StructType inputSchema;
+        String[] columnNamesArray;
+        List<String> columnNames;
+
+        List<VariableReferenceIterator> variableAccessExpressions = new ArrayList<>();
+        for (int expressionIndex = 0; expressionIndex < _expressions.size(); expressionIndex++) {
+            GroupByClauseSparkIteratorExpression expression = _expressions.get(expressionIndex);
+            inputSchema = df.schema();
```

Comment:
I can live with this because of "correctness first" -- could you make a GitHub issue for the future that it would be a nice performance improvement to be able to add several variables in one shot (for group by and let clauses)?

Reply:
Issue created